Sort-of rebase of 83: traj action aborts on error

gavanderhoorn opened the pull request:
Sort-of, as it's actually a bit more than what was included in #83.

The proposed changes make the `joint_trajectory_action` aware of the state of the (OEM) motion server program (as far as it publishes that in the form of `RobotStatus` messages). In two cases is this information taken into account:

 1. while accepting new goals
 2. while executing already accepted goals

In both cases, the following is monitored / checked:

 1. is an e-stop active?
 2. is there some other error active?
 3. does the motion server report `motion_possible` is `true`?

If any of these are not as expected, the goal is rejected, or, if a goal is currently being processed (ie: motion is execution), the goal is aborted and trajectory execution cancelled (as that should now work, with #260 merged (which still requires a cooperating driver of course)).

In all cases, the action result would be `INVALID_GOAL`, as unfortunately the `FollowJointTrajectory` action message does not define any really more suitable error codes.

Also in all cases: the reason for aborting or rejecting a goal is described in a human readable message.

When rejecting goals:

```
[ERROR] [...]: Rejecting goal: controller reported motion not possible (no further information)
[ERROR] [...]: Rejecting goal: controller reported (active) error (OEM code: 11003)
[ERROR] [...]: Rejecting goal: controller reported e-stop
```

When aborting goals:

```
[ERROR] [...]: Aborting goal: controller reported motion not possible (no further information)
[ERROR] [...]: Aborting goal: controller reported (active) error (OEM code: 11003)
[ERROR] [...]: Aborting goal: controller reported e-stop
```

(this is output from `fanuc_driver`: `11003` is *Deadman switch released*, which will indeed prevent any motion in manual mode)

While this is certainly not perfect (it doesn't address #265 nor #118), together with #263 and #260 and an (OEM) relay which behaves according to the spec, the UX of `industrial_robot_client` is improved quite a bit.

Lastly: as previous versions of the JTA did not check anything, misbehaving drivers would still be able to execute motions, even if they didn't properly populate all fields of the `RobotStatus` messages. In order to still allow usage of such systems, two new parameters were added which can be used to revert the JTA back to its previous behaviour (ie: don't check anything, always forward goals to drivers).

In the current implementation, the default is to disable these new checks. Something to discuss is whether we should make the default to *enable* the new checks, and users with misbehaving drivers should instead be required to explicitly enable the bw-compatible behaviour.

gavanderhoorn committed:
jta: abort goal on e-stop or error

gavanderhoorn committed:
traj_action: abort/reject based on motion server state

The action server will now take (OEM) motion server state into account when accepting or executing goals.

If the motion server reports (via the state server) it cannot currently execute motion, either because of a controller error or e-stop, or because a required program is not running on the (OEM) controller, new goals are rejected and currently executing goals are cancelled.

Cancelled trajectories should not resume executing once the (OEM) controller error state is cleared. This will however depend on a cooperating motion server (ie: it must manage its queues properly and watch for and act on errors on the OEM controller).

Cancellations and rejections are accompanied by appropriate human error messages, giving the reason for why a goal was cancelled or rejected. Unfortunately this is mostly only human readable as the FollowJointTrajectory action doesn't currently define appropriately expressive result codes.

gavanderhoorn committed:
traj_action: support ignoring motion server state

To keep the behaviour of the action server bw-compatible with the previous behaviour, add two ROS parameters which allow to:

 - ignore motion server state completely
 - treat UNKNOWN motion server state as ON or OFF (where appropriate)

By default motion server state is neither ignored nor are UNKNOWNs acceptable values (and so UNKNOWNs will result in rejected goals and aborted trajectory executions).

gavanderhoorn committed:
traj_action: use TriState convenience functions

Instead of manually comparing against TriState enum values.

gavanderhoorn committed:
traj_action: silence within-goal-constraints check

It doesn't need to be *this* verbose.

gavanderhoorn committed:
traj_action: formatting

gavanderhoorn committed:
traj_action: move is_*(..) to utils

gavanderhoorn committed:
robot_client: configure new params

But keep previous behaviour.

gavanderhoorn committed:
traj_action: split cancelling goal and stopping relay

So we can reuse stopping the relay.

gavanderhoorn commented:
Note: 55034f5 is basically what #83 proposed, but it also actually signals to the relay it should stop sending points.

Subsequent commits rewrite this into what the current implementation does.

gavanderhoorn commented:
I've requested @Levi-Armstrong and/or @JeremyZoss review this, as they've been involved with the development of `industrial_core` in the past.

@simonschmeisser: if you're still using this, and have some time, I'd appreciate your opinion as well.

Note: I do not intent to keep developing all of this. I just wanted to make sure I can get #83 in before releasing a new version of `industrial_core` into Melodic and Noetic. But since #83 actually only did half of what it was supposed to, I opted to flesh it out a bit more. And with #260 in place, it was actually not *that* difficult.

Note 2: it's still possible for trajectory relays to keep moving after an e-stop or error state is cleared: there is no way for the ROS side here (so the IRC node) to enforce anything. The streamer will now in all cases send a "traj abort" message, but if the trajectory and state relays (running on the OEM controller) do not cooperate, there isn't much we can do.

`fanuc_driver` fi will continue to execute the last sent point after resuming the relay program. Then it will stop. `fanuc_driver_exp` is slightly better, in that it purges its internal motion buffer and actually `SKIP`s the current segment. Other drivers may behave similarly (`abb_driver` fi).

gavanderhoorn commented:
This probably also addresses #131 partly, as this should not be possible any more (the JTA should reject the goal due to the controller not being in AUTO mode, or at least not reporting that `motion_possible` is `true`):

> This happens for me ie when the controller is set to "Manual mode" instead of "automatic". The mitsubishi controller will allow for upload but returns an error once I try to start execution of the trajectory. 

Of course it will depend on the (OEM) state program to properly report controller state (ie: `motion_possible` should accurately reflect whether motion is possible or not).

gavanderhoorn commented:
Friendly ping.

I don't believe this needs extensive testing.

The main question is whether we want to change the default behaviour (from not checking to checking).

simonschmeisser commented on industrial_robot_client/include/industrial_robot_client/joint_trajectory_action.h:
```diff
@@ -176,6 +176,33 @@ class JointTrajectoryAction
+   * cannot accurately report controller status (ie: have UNKNOWNs in
+   * the RobotStatus messages they publish).
+   */
+  bool ignore_motion_server_error_;
```
```suggestion
  bool ignore_motion_server_error_ = false;
```
I think it's already implicitly initialized to that but I'd prefer having this explicitely

simonschmeisser commented on industrial_robot_client/include/industrial_robot_client/joint_trajectory_action.h:
```diff
@@ -176,6 +176,33 @@ class JointTrajectoryAction
+   * are unable to accurately report controller status, by allowing
+   * the action server to assume UNKNOWN == OK.
+   */
+  bool consider_status_unknowns_ok_;
```
```suggestion
  bool consider_status_unknowns_ok_ = true;
```

simonschmeisser commented on industrial_robot_client/launch/robot_interface_download.launch:
```diff
@@ -30,6 +30,9 @@
   <!-- joint_trajectory_action: provides actionlib interface for high-level robot control -->
-  <node pkg="industrial_robot_client" type="joint_trajectory_action" name="joint_trajectory_action"/>
+  <node pkg="industrial_robot_client" type="joint_trajectory_action" name="joint_trajectory_action">
+    <param name="ignore_motion_server_error" type="bool" value="true" />
```
```suggestion
    <param name="ignore_motion_server_error" type="bool" value="false" />
```
I really appreciate this being tunable but I think we should be strict by default as it can introduce surprising delayed movements if you later enable movements (maybe  ... maybe not ... ?)

gavanderhoorn replied:
So this is indeed the main point of discussion here.

Doing this by default has the potential to break ppl's setups.

We could merge this PR as-is and not enable it on Melodic, then branch for Noetic and enable it there by default.

I hate branching for these little things, but I don't really see another option.

simonschmeisser replied:
If we leave it disabled, how would you enable it manually? copy paste the launch file?

I totally agree on avoiding branch-mess where possible

gavanderhoorn replied:
I see now that I forgot to export the `param`s using `arg`s. I'll add that. That would allow overriding the defaults.

gavanderhoorn replied:
Seems I've forgotten to fix the `arg`s missing. I'll need to address that before merging this.

simonschmeisser commented on industrial_robot_client/launch/robot_interface_streaming.launch:
```diff
@@ -30,7 +30,9 @@
   <!-- joint_trajectory_action: provides actionlib interface for high-level robot control -->
-  <node pkg="industrial_robot_client" type="joint_trajectory_action" name="joint_trajectory_action"/>
+  <node pkg="industrial_robot_client" type="joint_trajectory_action" name="joint_trajectory_action">
+    <param name="ignore_motion_server_error" type="bool" value="true" />
```
```suggestion
    <param name="ignore_motion_server_error" type="bool" value="false" />
```

simonschmeisser commented on industrial_robot_client/src/joint_trajectory_action.cpp:
```diff
@@ -45,12 +45,28 @@ const double JointTrajectoryAction::DEFAULT_GOAL_THRESHOLD_ = 0.01;
                    boost::bind(&JointTrajectoryAction::cancelCB, this, _1), false), has_active_goal_(false),
-                   controller_alive_(false), has_moved_once_(false), name_("joint_trajectory_action")
+                   controller_alive_(false), has_moved_once_(false), name_("joint_trajectory_action"),
+                   ignore_motion_server_error_(false), consider_status_unknowns_ok_(false)
```
```suggestion
                   ignore_motion_server_error_(false), consider_status_unknowns_ok_(true)
```
It's a bit too strict like that

I didn't see the initialization here on first glance, doing it here or in the header is a mater of preference I guess

gavanderhoorn replied:
setting this to `true` does decrease the value of the checks.

Ideally this would be a per-field setting, but that was a bit of a bridge-too-far for me when I wrote this.

simonschmeisser replied:
Yes, it's a bit less strict that way but also less likely to actually break setups (?). Hard to know without looking up all the drivers out there

simonschmeisser commented on industrial_robot_client/src/joint_trajectory_action.cpp:
```diff
@@ -45,12 +45,28 @@ const double JointTrajectoryAction::DEFAULT_GOAL_THRESHOLD_ = 0.01;
 
+  // Two parameters for bw-compatibility with the 'old' behaviour.
+  // Setting these by default to TRUE, to maintain the previous behaviour
+  pn.param("ignore_motion_server_error", ignore_motion_server_error_, true);
```
```suggestion
  pn.param("ignore_motion_server_error", ignore_motion_server_error_, ignore_motion_server_error_);
```
again a mater of preference I guess but it reduces the number of places where you can change the default

simonschmeisser commented on industrial_robot_client/src/joint_trajectory_action.cpp:
```diff
@@ -45,12 +45,28 @@ const double JointTrajectoryAction::DEFAULT_GOAL_THRESHOLD_ = 0.01;
+  // Two parameters for bw-compatibility with the 'old' behaviour.
+  // Setting these by default to TRUE, to maintain the previous behaviour
+  pn.param("ignore_motion_server_error", ignore_motion_server_error_, true);
+  pn.param("consider_status_unknowns_ok", consider_status_unknowns_ok_, true);
```
```suggestion
  pn.param("consider_status_unknowns_ok", consider_status_unknowns_ok_, consider_status_unknowns_ok_);
```

simonschmeisser commented on industrial_robot_client/src/joint_trajectory_action.cpp:
```diff
@@ -123,6 +191,24 @@ void JointTrajectoryAction::goalCB(JointTractoryActionServer::GoalHandle gh)
 
+  // check robot can actually execute trajectory, if not, refuse goal.
+  // no point in accepting the goal only to cancel it immediately later
+  if (!is_motion_server_ok(last_robot_status_, consider_status_unknowns_ok_) && !ignore_motion_server_error_)
```
micro optimization ahead: check the bool first?

gavanderhoorn replied:
I think I'll leave it like this. It's indeed a very minor optimisation, and the current order reads almost like plain English: "if the motion server is not ok AND we don't ignore motion server errors, ...".

simonschmeisser commented on industrial_robot_client/src/joint_trajectory_action.cpp:
```diff
@@ -123,6 +191,24 @@ void JointTrajectoryAction::goalCB(JointTractoryActionServer::GoalHandle gh)
+    control_msgs::FollowJointTrajectoryResult rslt;
+    // the goal is actually probably OK, but we have to choose one of the existing
+    // constants, and this one comes closest
+    rslt.error_code = control_msgs::FollowJointTrajectoryResult::INVALID_GOAL;
```
we could also set the `rslt.error_string` ... not sure if anybody displays that though ... but I could imagine implementing that at some point in MoveIt

gavanderhoorn replied:
`setRejected(..)` does that for us: [api doc](https://docs.ros.org/en/melodic/api/actionlib/html/classactionlib_1_1ServerGoalHandle.html#af0a4d2debf5b19d985553903ddaef92c).

gavanderhoorn committed:
Make param read DRY (ignore srvr error)

Co-authored-by: Simon Schmeisser <s.schmeisser@gmx.net>

gavanderhoorn committed:
Make param read DRY (unknowns OK)

Co-authored-by: Simon Schmeisser <s.schmeisser@gmx.net>

gavanderhoorn commented:
@simonschmeisser: thanks for the review.

Would you see an opportunity to test this on any real hw?

I'd be interested to know whether this mitigates what you reported in #131 for instance.

simonschmeisser commented:
I'll try to fit in testing that on our Mitsubishi in the evening

gavanderhoorn committed:
traj_action: use camelCase, not snake_case

Follow the style of IRC.

gavanderhoorn committed:
traj_action: put TriState helpers in ns

isOn(..), isOff(..) and isUnknown(..) are a bit too generic to not isolate them a bit.

gavanderhoorn commented:
I actually don't have access to an *downloading* driver / robot, so I'd be interested to know whether the proposed changes actually work there.

If they do, we'll need to update the `.launch` file for download drivers as well.

simonschmeisser commented:
Thanks for pointing that out, I actually only have access to a _downloading_ robot/driver

gavanderhoorn commented:
> Thanks for pointing that out, I actually only have access to a _downloading_ robot/driver

I expect the changes to _accepting goals_ to work for both. I believe that should help with #131.

Aborting active goals will work, but whether the OEM server program also cooperates I can't say -- and will depend on its implementation of course.

gavanderhoorn committed:
traj_action: init new members where declared

Instead of as part of the ctor.

gavanderhoorn committed:
traj_action: clarify ctor overrides param values

gavanderhoorn committed:
traj_action: can only set params to defaults

We don't hard-code them in the ctor again.

gavanderhoorn committed:
traj_action: expose new params as args in the launch file

So their values can be changed by users.

gavanderhoorn committed:
traj_action: explain what the new launch args are for

gavanderhoorn commented:
Ok, so I've changed things around a bit:

 - defaults are only set once per context (in the JTA and as parameters in the `.launch` file)
 - by default everything is enabled: errors are not ignored, neither are UNKNOWNs treated as OK
 - `robot_interface_streaming.launch` exposes two new `arg`s, which allow changing the two bw-compatibility knobs introduced

We can still decide to set `consider_status_unknowns_ok` to `true` by default.

We could also say that the default should be `false`, and drivers which can't function that way should explicitly override the default `arg` and set it to `true`.

gavanderhoorn commented:
@simonschmeisser: any luck? :)

gavanderhoorn commented:
@simonschmeisser: *very* friendly ping.

simonschmeisser commented:
Sorry, I'm on the road (well tracks to be exact ...) but will try to give this some time this Friday

gavanderhoorn commented:
@simonschmeisser: friendly reminder.

(only doing this as I tend to forget when I sort-of promise someone I would do something three days later while sitting in a train ;) )

simonschmeisser commented:
The immediate feedback when being in manual mode/e-stop/drives not powered is quite nice.

```
[ERROR] [1625839551.379974859]: Rejecting goal: controller reported motion not possible (no further information)
[ WARN] [1625839551.380040937]: Controller  failed with error INVALID_GOAL: 
[ WARN] [1625839551.380065314]: Controller handle  reports status FAILED
[ INFO] [1625839551.380084931]: Completed trajectory execution with status FAILED ...
[ERROR] [1625839551.380121015]: Error executing plan:  "FAILED"
```
I wonder if this could or should be more specific?

Edit: e-stop is printed out correctly:
```
[ERROR] [1625840349.974565108]: Rejecting goal: controller reported e-stop

```

Otherwise I could not spot any regression with our mitsubishi downloading driver

gavanderhoorn commented:
> I wonder if this could or should be more specific?

well, we could inspect fields like `drives_powered` etc, but a) this was already enough work to implement and test (for a package basically in maintenance mode) and b) it gets quite confusing with drivers publishing `UNKNOWN`s.

Easy enough to do though: update `describeRobotStatusMsg(..)` in `industrial_robot_client/src/joint_trajectory_action.cpp` ([here](https://github.com/ros-industrial/industrial_core/pull/271/files#diff-ddfb1b2e4f9a12d814d443f1c6ecabbca5d682a6044344234c47b18e730776e5R137)).

Right now it singles out e-stops, then checks for any other errors (with potentially an OEM code) and finally it just checks `motion_possible`.

I'm not sure we can say for all supported systems that fi `drives_powered` is a requirement for `motion_possible` -- some robots have an "eco mode", which powers down the drives when they are not used for some time (even with a program running), and will automatically enable them again when needed.

These sort of things made me hesitant to check anything more than `motion_possible`, as at least for that field, we can rely on the "insight" and understanding of the driver developer to figure out whether motion would be possible and report it accordingly.

I would probably recommend applications (ie: other nodes) to subscribe to `robot_status` and inspect the messages if further information could help determine a course of action. The IRC doesn't have any way to communicate it, as we always only use `INVALID_GOAL`.

gavanderhoorn commented:
Just noticed this in your comment:

```
[ WARN] [1625839551.380040937]: Controller  failed with error INVALID_GOAL: 
```

I would have expected the message passed to `setRejected(..)` to show up after `INVALID_GOAL:`, but that doesn't appear to be the case.

To make the formatting for empty controller names somewhat easier to read, I've submitted https://github.com/ros-planning/moveit/pull/2761.

simonschmeisser commented:
Yes, fully agree.

One thing I was worried about is that this PR would reject goals in manual mode but it doesn't and that's a good decision as well. We show a confirmation dialog when in manual mode or anything else considered erroneous in our software and it can get annoying during initial bring up. So it's good that we don't need to change parameters to allow movements in manual mode.

gavanderhoorn commented:
> One thing I was worried about is that this PR would reject goals in manual mode but it doesn't and that's a good decision as well.

well, I didn't do anything special for that. It's all part of "only look at `motion_possible`".

If the driver reports motion should be possible, we're not going to doubt that by looking at other fields and then trying to be clever.

gavanderhoorn commented:
What we *could* do, is only include `drives_powered` et al. in the description (ie: the error message), but not in the check.

But I'm not sure that would work either, as it may lead to the wrong conclusion *on the human side* when they read the error message.

gavanderhoorn commented on industrial_robot_client/include/industrial_robot_client/utils.h:
```diff
@@ -36,6 +36,9 @@
 #include <string>
 #include <map>
 
+#include <industrial_msgs/RobotStatus.h>
```
```suggestion
#include <industrial_msgs/TriState.h>
```

gavanderhoorn committed:
We only need TriState, not the full RobotStatus

simonschmeisser commented:
We would need to set the `error_text` string field of the `FollowJointTrajectory` msg for it to be printed there

I'm not sure why we have the empty controller name in all of our controller files ... what do you usually put there? Something like `mitsubishi driver`?

gavanderhoorn commented:
> We would need to set the `error_text` string field of the `FollowJointTrajectory` msg for it to be printed there

According to [the api doc](https://docs.ros.org/en/melodic/api/actionlib/html/classactionlib_1_1ServerGoalHandle.html#af0a4d2debf5b19d985553903ddaef92c) (and the sources), anything you pass to `setRejected(..)` gets assigned to the `error_text`. At least that's how I interpreted that.

> I'm not sure why we have the empty controller name in all of our controller files ... what do you usually put there? Something like `mitsubishi driver`?

It's most likely empty because the template [in the tutorial lots of people refer to](http://wiki.ros.org/Industrial/Tutorials/Create_a_MoveIt_Pkg_for_an_Industrial_Robot#Update_Configuration_Files) has it set to `""`.

If you set it to something non-empty, MoveIt will start looking in a different place for your action topics. That's not necessarily a problem though, you'd just have to update your driver to offer the action server there (or `remap`).

simonschmeisser commented:
> According to [the api doc](https://docs.ros.org/en/melodic/api/actionlib/html/classactionlib_1_1ServerGoalHandle.html#af0a4d2debf5b19d985553903ddaef92c) (and the sources), anything you pass to `setRejected(..)` gets assigned to the `error_text`. At least that's how I interpreted that.

But we have two text variables for the error here. The generic one you are setting and the redundant one in http://docs.ros.org/en/api/control_msgs/html/action/FollowJointTrajectory.html

gavanderhoorn commented:
O, crap, you're right.

I'd somehow confused those.

Perhaps MoveIt should check both and display them.

gavanderhoorn commented on industrial_robot_client/src/joint_trajectory_action.cpp:
```diff
@@ -123,6 +189,24 @@ void JointTrajectoryAction::goalCB(JointTractoryActionServer::GoalHandle gh)
+    control_msgs::FollowJointTrajectoryResult rslt;
+    // the goal is actually probably OK, but we have to choose one of the existing
+    // constants, and this one comes closest
+    rslt.error_code = control_msgs::FollowJointTrajectoryResult::INVALID_GOAL;
```
```suggestion
    rslt.error_code = control_msgs::FollowJointTrajectoryResult::INVALID_GOAL;
    rslt.error_string = reject_msg;
```

gavanderhoorn commented on industrial_robot_client/src/joint_trajectory_action.cpp:
```diff
@@ -227,9 +309,32 @@ void JointTrajectoryAction::controllerStateCB(const control_msgs::FollowJointTra
+    control_msgs::FollowJointTrajectoryResult result;
+    // would like to use a better error constant, but we have to choose one of the existing
+    // constants, and this one comes closest
+    result.error_code = control_msgs::FollowJointTrajectoryResult::INVALID_GOAL;
```
```suggestion
    result.error_code = control_msgs::FollowJointTrajectoryResult::INVALID_GOAL;
    result.error_string = abort_msg;
```

gavanderhoorn committed:
Send abort_msg back to client

gavanderhoorn committed:
Send reject_msg back to client

gavanderhoorn commented:
So with the technical side out of the way, the question becomes: do we want to merge this into `melodic-devel` and change the default behaviour?

@simonschmeisser @Levi-Armstrong @anyone-with-an-opinion-really?

gavanderhoorn commented:
With 0eeb23f0ff387a69deecaab66e9106e2f7af3228 and 0fa8bba32bbc1fc2a2e0bc140119ac4ab6e33817, MoveIt now reports something like:

```
[ WARN] [1626612647.088687582]: Controller '' failed with error INVALID_GOAL: Rejecting goal: controller reported motion not possible (no further information)
[ WARN] [1626612647.088809223]: Controller handle '' reports status FAILED
[ INFO] [1626612647.088865293]: Completed trajectory execution with status FAILED ...
[ INFO] [1626612647.088960767]: Execution completed: FAILED
```

The double *controller* there can be a bit confusing.

I was thinking of maybe adding "OEM" or "robot" there, so we would get:

```
[ WARN] [1626612647.088687582]: Controller '' failed with error INVALID_GOAL: Rejecting goal: OEM controller reported motion not possible (no further information)
```

But this would of course not be accurate if the server side would not run on an OEM controller.

simonschmeisser commented:
I'm not sure how many people understand what you mean by `OEM` so I'd rather go for `robot controller`

gavanderhoorn commented on industrial_robot_client/src/joint_trajectory_action.cpp:
```diff
@@ -107,6 +121,58 @@ void JointTrajectoryAction::watchdog(const ros::TimerEvent &e)
+  if (utils::tri_state::isOn(msg->e_stopped, unknown_is_on))
+  {
+    ss.clear();
+    ss << "controller reported e-stop";
```
```suggestion
    ss << "robot controller reported e-stop";
```

gavanderhoorn commented on industrial_robot_client/src/joint_trajectory_action.cpp:
```diff
@@ -107,6 +121,58 @@ void JointTrajectoryAction::watchdog(const ros::TimerEvent &e)
+  else if (msg->error_code != 0 || utils::tri_state::isOn(msg->in_error, unknown_is_on))
+  {
+    ss.clear();
+    ss << "controller reported (active) error";
```
```suggestion
    ss << "robot controller reported (active) error";
```

gavanderhoorn commented on industrial_robot_client/src/joint_trajectory_action.cpp:
```diff
@@ -107,6 +121,58 @@ void JointTrajectoryAction::watchdog(const ros::TimerEvent &e)
+  else if (utils::tri_state::isOff(msg->motion_possible, unknown_is_on))
+  {
+    ss.clear();
+    ss << "controller reported motion not possible (no further information)";
```
```suggestion
    ss << "robot controller reported motion not possible (no further information)";
```

gavanderhoorn committed:
Be clear about which controller reported a problem

gavanderhoorn committed:
Be clear about which controller reported a problem

gavanderhoorn committed:
Be clear about which controller reported a problem

simonschmeisser reviewed (dismissed):
There don't seem to be that many opinionated people around any more ...

I vote for merging this as is to `melodic-devel` as I see the slight change in behavior as a bug fix basically

gavanderhoorn committed:
traj_action: also expose params as args in download launch file

So their values can be changed by users.

gavanderhoorn commented:
Ok, to move this forward, I'm going to create a "post noetic" branch.

On *that* branch, everything discussed here will default to `false` (ie: `ignore_motion_server_error` and `consider_status_unknowns_ok`).

On both Melodic and Noetic, they'll default to `true` (to maintain the current behaviour).

People still using `industrial_robot_client` and likely to keep use it could build from source. The binary releases will just play-nice.

simonschmeisser commented:
Friendly ping, could be part of ros-o
